app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -9,71 +12,109 @@ class OmniauthCallbacksController < Devise::OmniauthCallbacksController
     # SAML doesn't support the CSRF protection
     protect_from_forgery except: %i[mocksaml bird nbp]
 
-    def sso_callback # rubocop:disable Metrics/AbcSize
-      # Check if an existing user is already signed in (passed through the RelayState)
-      # and trying to add a new identity to their account. If so, we load the user information
-      # and set it as the current user. This is necessary to avoid creating a new user.
-      current_user = User.find_by(id: OmniAuth::NonceStore.pop(params[:RelayState]))
-
-      # For existing users, we want to redirect to their profile page after adding a new identity
-      store_location_for(:user, edit_user_registration_path) if current_user.present?
-
-      # The instance variable `@user` is used by Devise internally and should be set here
-      @user = User.from_omniauth(request.env['omniauth.auth'], current_user)
-
-      if @user.persisted?
-        # The `sign_in_and_redirect` will only proceed with the login if the account has been confirmed
-        # either through trust by the IdP or with a confirmation mail. Until then, a flash message
-        # notifies users about the missing email confirmation
-        sign_in_and_redirect @user, event: :authentication
-        # We use the configured OmniAuth camilization to include the user-facing name of the provider in the flash message
-        set_flash_message(:notice, :success, kind: OmniAuth::Utils.camelize(provider)) if is_navigational_format?
-        session['omniauth_provider'] = provider
+    def sso_callback
+      # `current_user` refers to an existing user signed in before starting the SAML workflow.
+      # Thus, this value is only present when adding a new identity to an authenticated account.
+      if current_user.present?
+        add_idp_to_existing_user
+      elsif user_identity.persisted?
+        sign_in_with_identity
       else
-        if is_navigational_format? && @user.errors.any?
-          # We show validation errors to the user, for example because required data from the IdP was missing
-          set_flash_message(:alert, :failure, kind: OmniAuth::Utils.camelize(provider),
-            reason: @user.errors.full_messages.join(', '))
-        end
-        # Removing extra as it can overflow some session stores
-        session["devise.#{provider}_data"] = request.env['omniauth.auth'].except('extra')
-        redirect_to new_user_registration_url
+        register_new_user
       end
     end
 
     alias bird sso_callback
     alias nbp sso_callback
     alias mocksaml sso_callback
 
-    def provider
-      request.env['omniauth.auth']&.provider || params[:provider]
-    end
+    def deauthorize # rubocop:disable Metrics/AbcSize
```

Comment:
I just made another observation: When deauthorizing the `nbp` connection, we are still keeping the `enmeshed` ID in the backend. Do you think that's okay or should we remove this one, too?

Reply:
No, I think we want to keep it. We are storing this for future use and we might need it even for users that deauthorized their nbp IdP.

Reply:
Okay, fine. Then, let's add a short comment to the code about that intention. 🙏 

I think that there is no dedicated handling needed when keeping the enmeshed ID. Deleting an account will still remove the identity, and I am currently not aware of any other code path that could attempt to re-add an identity with the same enmeshed ID: Deleting a contact in the app is currently not possible and re-adding the same contact is neither. Therefore, one would need to sign up for CodeHarbor again and use a new profile in the enmeshed app, which will also generate a new enmeshed ID.